R/filter_state_api_NEW.R

```diff
@@ -0,0 +1,350 @@
+
+all_filters <- filter_settings(
+  filter_one,
+  filter_two,
+  filter_three,
+  exclude = list(
+    "dataname1" = "varname2"
+  )
+)
+# all_filters
+# c(all_filters[1], all_filters[2])
+
+
+# get slices where feature is value
+extract_by_feat <- function(tss, feature, value) {
+  checkmate::assert_class(tss, "teal_slices")
+  Filter(function(x) x[[feature]] == value, tss)
+}
+extract_by_feat(all_filters, "dataname", "dataname1")
+
+# get feature from all slices
+extract_feat <- function(tss, feature) {
+  checkmate::assert_class(tss, "teal_slices")
+  lapply(tss, function(x) x[[feature]])
+}
+extract_feat(all_filters, "dataname")
+
+# get slices where logical predicate is TRUE
+extract_fun <- function(tss, expr) {
+  checkmate::assert_class(tss, "teal_slices")
+  expr <- substitute(expr)
+  checkmate::assert_class(expr, "call")
+  Filter(function(x) isTRUE(eval(expr, x)), tss)
+}
+extract_fun(all_filters, dataname == "dataname2")
+extract_fun(all_filters, extras$extra2 == "extratwo")
+
+# string version
+extract_fun_s <- function(tss, expr) {
+  checkmate::assert_class(tss, "teal_slices")
+  checkmate::assert_string(expr)
+  expr <- str2lang(expr)
+  Filter(function(x) isTRUE(eval(expr, x)), tss)
+}
+extract_fun_s(all_filters, 'dataname == "dataname2"')
+x = "dataname2"
+extract_fun_s(all_filters, sprintf('dataname == "%s"', x))
+
+# add elements to extras in all slices
+add_extras <- function(tss, extras) {
+  ans <- lapply(tss, function(ts) {
+    ts$extras <- c(ts$extras, extras)
+    ts
+  })
+  attributes(ans) <- attributes(tss)
+  ans
+}
+
+# convert list to teal_slice
+as.teal_slice <- function(x) {
+  do.call(filter_var, x)
+}
+
+# # convert list to teal_slice
+# as.teal_slices <- function(x, master) {
+#   varnames <- lapply(x, names)
+#   selecteds <- unlist(lapply(x, function(x) lapply(x, function(x) x$selected)), recursive = FALSE)
+#   keep_nas <- unlist(lapply(x, function(x) lapply(x, function(x) x$keep_na)), recursive = FALSE)
+#   keep_infs <- unlist(lapply(x, function(x) lapply(x, function(x) x$keep_inf)), recursive = FALSE)
+#
+#   datanames <- character(0)
+#   for (d in seq_along(x)) {
+#     for (v in seq_along(x[[d]])) {
+#       datanames <- c(datanames, names(x)[d])
+#     }
+#   }
+#   varnames <- unname(unlist(lapply(x, names)))
+#
+#   slices <- .mapply(
+#     FUN = filter_var,
+#     dots = list(
+#       dataname = datanames,
+#       varname = varnames,
+#       selected = selecteds,
+#       keep_na = keep_nas,
+#       keep_inf = keep_infs
+#     ),
+#     MoreArgs = list(
+#       choices = NULL,
+#       fixed = FALSE
+#     )
+#   )
+#
+#   if (!missing(master)) {
+#     slices <- lapply(slices, function(x) {
+#       x$dataname <- master
+#       x
+#     })
+#   }
+#
+#   do.call(filter_settings, as.list(slices, list(exclude = list(), count_types = "none")))
+# }
+as.teal_slices <- function(x) {
```

Comment:
very good idea to simplify the code in the `FilteredData` class and make it easy to test separately!

Reply:
I was thinking to export this function but It will be only temporary. Please remove `@export` tag